### PR TITLE
made integrity hash to be generated based on lockfile entries, not lockfile content

### PR DESCRIPTION
**Pull request by bestander**
**Summary**
This will fix some of the cases in #569
- Integrity generation in `check` command was based on lockfile on disk content rather than values inside lockfiles.
- Integrity check in `install.js` command can be done after dependencies are resolved, i.e. lock file is read and parsed.

**Test plan**

Added test.

Manual test plan:
```
bestander@DESKTOP-9BB67ID:/mnt/c/Users/besta/work/temp$ ../yarn/bin/yarn install
yarn install v0.17.0-0
warning No license field
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Building fresh packages...
success Saved lockfile.
Done in 0.28s.
bestander@DESKTOP-9BB67ID:/mnt/c/Users/besta/work/temp$ ../yarn/bin/yarn install
yarn install v0.17.0-0
warning No license field
[1/4] Resolving packages...
success Already up-to-date.
Done in 0.12s.
bestander@DESKTOP-9BB67ID:/mnt/c/Users/besta/work/temp$ ../yarn/bin/yarn install --force
yarn install v0.17.0-0
warning No license field
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Rebuilding all packages...
success Saved lockfile.
Done in 0.29s.

bestander@DESKTOP-9BB67ID:/mnt/c/Users/besta/work/temp$ ../yarn/bin/yarn check --integrity
yarn check v0.17.0-0
warning No license field
success Folder in sync.
Done in 0.12s.
bestander@DESKTOP-9BB67ID:/mnt/c/Users/besta/work/temp$ rm node_modules/mime-db/package.json
bestander@DESKTOP-9BB67ID:/mnt/c/Users/besta/work/temp$ ../yarn/bin/yarn check --integrity
yarn check v0.17.0-0
warning No license field
success Folder in sync.
Done in 0.12s.
bestander@DESKTOP-9BB67ID:/mnt/c/Users/besta/work/temp$ ../yarn/bin/yarn check
yarn check v0.17.0-0
warning No license field
error mime-types#mime-db not installed
error Found 1 errors.
info Visit https://yarnpkg.com/en/docs/cli/check for documentation about this command.
```

### CODE REVIEW

File: src/cli/commands/install.js [hunk 1]
```diff
@@ -524,7 +524,8 @@ export class Install {
       };
     }
 
-    const actual = this.generateIntegrityHash(this.lockfile.source, patterns);
+    const lockSource = lockStringify(this.lockfile.getLockfile(this.resolver.patterns));
```
Comment (bestander) [inline src/cli/commands/install.js]: This is wrong, this.resolver.patterns may not be defined yet

**Comment by bestander** [comment]
Ready for review

**Comment by kittens** [comment]
Do you have any stats on the performance implications of this change? We've been praised a lot on our super quick execution when `node_modules` is up to date and this changes that and will be significantly slower.

**Comment by bestander** [comment]
Cold install should take the same time, I just moved the bailout down the chain.
For hot install I can measure react-native repo before/after, anything else?

**Comment by bestander** [comment]
Test has been run on my battery powered Surface Book i5 on linux.
Used on React Native github repo.

Current stable Yarn 0.16.1, average ~0.5 seconds
```
bestander@DESKTOP-9BB67ID:/mnt/c/Users/besta/work/react-native$ time yarn install
yarn install v0.16.1
success Already up-to-date.
Done in 0.23s.

real    0m0.558s
user    0m0.438s
sys     0m0.141s
bestander@DESKTOP-9BB67ID:/mnt/c/Users/besta/work/react-native$ time yarn install
yarn install v0.16.1
success Already up-to-date.
Done in 0.22s.

real    0m0.546s
user    0m0.422s
sys     0m0.188s
bestander@DESKTOP-9BB67ID:/mnt/c/Users/besta/work/react-native$ time yarn install
yarn install v0.16.1
success Already up-to-date.
Done in 0.21s.

real    0m0.474s
user    0m0.359s
sys     0m0.172s
bestander@DESKTOP-9BB67ID:/mnt/c/Users/besta/work/react-native$ time yarn install
yarn install v0.16.1
success Already up-to-date.
Done in 0.21s.

real    0m0.476s
user    0m0.375s
sys     0m0.172s
bestander@DESKTOP-9BB67ID:/mnt/c/Users/besta/work/react-native$ time yarn install
yarn install v0.16.1
success Already up-to-date.
Done in 0.22s.

real    0m0.535s
user    0m0.391s
sys     0m0.203s
```

With this change average 0.95 seconds
```
bestander@DESKTOP-9BB67ID:/mnt/c/Users/besta/work/react-native$ time ../yarn/bin/yarn install
yarn install v0.17.0-0
[1/4] Resolving packages...
success Already up-to-date.
Done in 0.67s.

real    0m1.012s
user    0m0.891s
sys     0m0.266s
bestander@DESKTOP-9BB67ID:/mnt/c/Users/besta/work/react-native$ time ../yarn/bin/yarn install
yarn install v0.17.0-0
[1/4] Resolving packages...
success Already up-to-date.
Done in 0.62s.

real    0m0.947s
user    0m0.766s
sys     0m0.281s
bestander@DESKTOP-9BB67ID:/mnt/c/Users/besta/work/react-native$ time ../yarn/bin/yarn install
yarn install v0.17.0-0
[1/4] Resolving packages...
success Already up-to-date.
Done in 0.62s.

real    0m0.941s
user    0m0.734s
sys     0m0.281s
bestander@DESKTOP-9BB67ID:/mnt/c/Users/besta/work/react-native$ time ../yarn/bin/yarn install
yarn install v0.17.0-0
[1/4] Resolving packages...
success Already up-to-date.
Done in 0.61s.

real    0m0.927s
user    0m0.906s
sys     0m0.188s
bestander@DESKTOP-9BB67ID:/mnt/c/Users/besta/work/react-native$ time ../yarn/bin/yarn install
yarn install v0.17.0-0
[1/4] Resolving packages...
success Already up-to-date.
Done in 0.62s.

real    0m0.951s
user    0m0.750s
sys     0m0.344s
```

**Comment by bestander** [comment]
On MacBook pro the times are 0.35 and 0.7 seconds for the same test.
cc @kittens

**Pull request merged**